src/schemathesis/protocols.py

```diff
@@ -0,0 +1,141 @@
+"""Protocols that describe how concrete implementations should behave."""
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, TypeVar
+
+import requests
+from hypothesis.strategies import SearchStrategy
+from typing_extensions import Protocol
+
+from .hooks import HookDispatcher
+from .stateful import StatefulTest
+from .types import Body, Cookies, FormData, Headers, PathParameters, Query
+from .utils import GenericResponse, WSGIResponse
+
+
+class CaseProtocol(Protocol):
+    endpoint: "EndpointProtocol"
+    path_parameters: Optional[PathParameters]
+    headers: Optional[Headers]
+    cookies: Optional[Cookies]
+    query: Optional[Query]
+    body: Optional[Body]
+    form_data: Optional[FormData]
```

Comment:
I think that attributes that contain actual data (e.g. `form_data`, `body`) should not be exposed in the protocol and actually used by the concrete implementation via some method (which is defined in the protocol)